File: transpiler/javatests/com/google/j2cl/integration/java/jsmethodoverride/Main.java
/*
 * Copyright 2017 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package jsmethodoverride;

import static com.google.j2cl.integration.testing.Asserts.assertEquals;
import static com.google.j2cl.integration.testing.Asserts.assertTrue;

import jsinterop.annotations.JsMethod;
import jsinterop.annotations.JsPackage;
import jsinterop.annotations.JsType;

public class Main {
  public static void main(String... args) {
    testJsMethodOverride();
    testJsMethodOverrideWithBridges();
  }

  static class A {
    @JsMethod(name = "y")
    public int x() {
      return 1;
    }
  }

  @JsType
  static class B extends A {
    @Override
    public int x() {
      return 2;
    }
  }

  @JsMethod
  public static native int callY(Object o);

  public static void testJsMethodOverride() {
    A a = new A();
    B b = new B();
    assertTrue(1 == callY(a));
    assertTrue(2 == callY(b));
  }

  @JsType
  static class Parent<T> {
    public String withoutOverrideRenamed(T t) {
      return "Parent.notRenamed";
    }

    public String withOverrideRenamed(T t) {
      return "Parent.toBeRenamed";
    }
  }

  @JsType
  static class Child extends Parent<String> {
    public String withoutOverrideRenamed(String s) {
      return "Child.notRenamed";
    }

    @JsMethod(name = "withNewName")
    public String withOverrideRenamed(String s) {
      return "Child.renamed";
    }
  }

  @JsType(isNative = true, name = "?", namespace = JsPackage.GLOBAL)
  interface AllMethods {
    String withoutOverrideRenamed(Object o);

    String withOverrideRenamed(Object o);

    String withNewName(Object o);
  }

  public static void testJsMethodOverrideWithBridges() {
    Parent<?> p = new Parent<Object>();
    assertEquals("Parent.notRenamed", ((Parent<Object>) p).withoutOverrideRenamed(new Object()));
    assertEquals("Parent.toBeRenamed", ((Parent<Object>) p).withOverrideRenamed(new Object()));
    p = new Child();
    assertEquals("Child.notRenamed", ((Parent<String>) p).withoutOverrideRenamed(""));
    assertEquals("Child.renamed", ((Parent<String>) p).withOverrideRenamed(""));

    Child c = new Child();
    assertEquals("Child.notRenamed", c.withoutOverrideRenamed(""));
    assertEquals("Child.renamed", c.withOverrideRenamed(""));

    AllMethods am = (AllMethods) c;
    assertEquals("Child.notRenamed", am.withoutOverrideRenamed(""));
    // Call through the bridged method
    assertEquals("Child.renamed", am.withOverrideRenamed(""));
    // Call directly through the new name
    assertEquals("Child.renamed", am.withNewName(""));
  }
}

File: transpiler/javatests/com/google/j2cl/integration/kotlin/jsmethodoverride/Main.kt
/*
 * Copyright 2017 Google Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package jsmethodoverride

import com.google.j2cl.integration.testing.Asserts.assertEquals
import com.google.j2cl.integration.testing.Asserts.assertTrue
import jsinterop.annotations.JsMethod
import jsinterop.annotations.JsPackage
import jsinterop.annotations.JsType

fun main(vararg args: String) {
  testJsMethodOverride()
  testJsMethodOverrideWithBridges()
}

open class A {
  @JsMethod(name = "y") open fun x(): Int = 1
}

@JsType
class B : A() {
  override fun x(): Int = 2
}

@JsMethod external fun callY(o: Any?): Int

fun testJsMethodOverride() {
  val a = A()
  val b = B()
  assertTrue(1 == callY(a))
  assertTrue(2 == callY(b))
}

@JsType
open class Parent<T> {
  open fun withoutOverrideRenamed(t: T): String = "Parent.notRenamed"

  open fun withOverrideRenamed(t: T): String = "Parent.toBeRenamed"
}

@JsType
class Child : Parent<String>() {
  override fun withoutOverrideRenamed(s: String): String = "Child.notRenamed"

  @JsMethod(name = "withNewName")
  override fun withOverrideRenamed(s: String): String = "Child.renamed"
}

@JsType(isNative = true, name = "?", namespace = JsPackage.GLOBAL)
interface AllMethods {
  fun withoutOverrideRenamed(o: Any?): String?

  fun withOverrideRenamed(o: Any?): String?

  fun withNewName(o: Any?): String?
}

@Suppress("UNCHECKED_CAST")
fun testJsMethodOverrideWithBridges() {
  var p: Parent<*> = Parent<Any>()
  assertEquals("Parent.notRenamed", (p as Parent<Any>).withoutOverrideRenamed(Any()))
  assertEquals("Parent.toBeRenamed", (p as Parent<Any>).withOverrideRenamed(Any()))
  p = Child()
  assertEquals("Child.notRenamed", (p as Parent<String>).withoutOverrideRenamed(""))
  assertEquals("Child.renamed", (p as Parent<String>).withOverrideRenamed(""))

  val c = Child()
  assertEquals("Child.notRenamed", c.withoutOverrideRenamed(""))
  assertEquals("Child.renamed", c.withOverrideRenamed(""))

  val am = c as AllMethods
  assertEquals("Child.notRenamed", am.withoutOverrideRenamed(""))
  // Call through the bridged method
  assertEquals("Child.renamed", am.withOverrideRenamed(""))
  // Call directly through the new name
  assertEquals("Child.renamed", am.withNewName(""))
}
